### PR TITLE
CLIP model.

### PR DESCRIPTION
**Pull request by b0nce**

### CODE REVIEW

File: oml/transforms/images/torchvision/transforms.py [hunk 1]
```diff
@@ -25,4 +25,17 @@ def get_normalisation_resize_torch(im_size: int, mean: TNormParam = MEAN, std: T
...
 
 
-__all__ = ["get_augs_torch", "get_normalisation_torch", "get_normalisation_resize_torch"]
+def get_clip_transforms(
+    mean: TNormParam = (0.48145466, 0.4578275, 0.40821073),
+    std: TNormParam = (0.26862954, 0.26130258, 0.27577711),
+) -> Compose:
+    return t.Compose(
+        [
+            t.Lambda(lambda img: img.convert("RGB") if img.mode != "RGB" else img),
```
Comment (AlekseySh) [inline oml/transforms/images/torchvision/transforms.py:34]: who do we need this converter? what is the assumption for the image file?

File: tests/test_oml/test_models/test_models_creation.py [hunk 2]
```diff
@@ -1,36 +1,50 @@
...
-    for key in constructor.pretrained_models.keys():
-        net = constructor(weights=key, arch=key.split("_")[0], **args)
-        net(torch.randn(1, 3, 224, 224))
+    if not isinstance(constructor, ViTCLIPExtractor):
+        for key in constructor.pretrained_models.keys():
+            net = constructor(weights=key, arch=key.split("_")[0], **args)
+            net(torch.randn(1, 3, 224, 224))
+    else:
+        """
+        We are skipping tests for OpenAI's CLIP due to its hardcoded cuda requirement in forward.
```
Comment (AlekseySh) [inline tests/test_oml/test_models/test_models_creation.py:44]: hm, let's discuss offline what can we do

File: oml/models/vit/clip.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+        return x
+
+
+class ViTCLIPExtractor(IExtractor):
+    pretrained_models: Dict[str, Any] = {}  # there are too many pretrained architectures, so we will skip them
+
+    def __init__(
+        self,
+        arch: Optional[str] = None,
```
Comment (AlekseySh) [inline oml/models/vit/clip.py:158]: why the arch is optional?

File: tests/test_oml/test_models/test_models_creation.py [hunk 4]
```diff
@@ -1,36 +1,50 @@
...
+vit_clip_args = {"normalise_features": False}
 resnet_args = {"normalise_features": True, "gem_p": 7.0, "remove_fc": False, "strict_load": True}
 
 
 @pytest.mark.parametrize(
-    "constructor,args,default_arch", [(ViTExtractor, vit_args, "vits16"), (ResnetExtractor, resnet_args, "resnet50")]
+    "constructor,args,default_arch",
+    [
+        (ViTExtractor, vit_args, "vits16"),
+        (ViTCLIPExtractor, vit_clip_args, None),
```
Comment (AlekseySh) [inline tests/test_oml/test_models/test_models_creation.py:21]: the same as the comment above: not sure about the purpose of having None as a valid architecture parameter

File: oml/models/vit/clip.py (hunk 3 above)
Comment (AlekseySh) [reply oml/models/vit/clip.py:158]: as for me, embed_dim, layers, width... are parameters related to the architecture. I like the way how we encapsulate them in the architecture arguments in our default ViT.

File: oml/models/vit/clip.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
+from collections import OrderedDict
```
Comment (AlekseySh) [inline oml/models/vit/clip.py:1]: please, add the comment with the link to the original source code
and let's keep the 3d party code and our code separately as we did it for ViT (there are 3 files under the vit directory, but only 1 of them contains our code)

File: oml/models/vit/clip.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+from typing import Any, Dict, Optional
+
+import torch
+from torch import nn
+
+from oml.interfaces.models import IExtractor
+from oml.models.utils import remove_prefix_from_state_dict
+from oml.utils.io import download_checkpoint
+
+CLIP_MODELS = {
```
Comment (AlekseySh) [inline oml/models/vit/clip.py:12]: can we follow the same structure as we have in ViT?

when we have a set of functions-constructors and the parameters related to the architecture exist only inside these functions

let's discuss it offline

File: oml/models/vit/clip.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+            weights = cfg["weights"]
+            self.visual = torch.jit.load(Path(weights), map_location="cpu").visual
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        with torch.autocast(x.device.type):
+            if not self.normalize:
+                return self.visual.forward(x)
+            else:
+                res = self.visual.forward(x)
+                return res / res.norm(dim=1, keepdim=True)
```
Comment (AlekseySh) [inline oml/models/vit/clip.py:218]: please, check how we normalise features in other models

**Review by AlekseySh — COMMENTED** [review]
I am afraid that this CLIP model may be hard to find for the user. The same for other pretrained models since in readme we have the links only to the models which we obtained by ourselves.

So, I guess it's a good time to introduce one more table in the Models Zoo with the scores of pretrained models on 4 public benchmarks. It will be useful for us and for users since they can see easily which models we have.

Could you add such a table to the readme? For CLIP, VIT and MoCo

File: oml/models/vit/clip.py (hunk 3 above)
Comment (AlekseySh) [reply oml/models/vit/clip.py:158]: let's discuss it offline

File: oml/models/vit/clip.py (hunk 6 above)
Comment (AlekseySh) [reply oml/models/vit/clip.py:12]: It would be also good to have the same `pretrained_models` field in that class

File: oml/models/vit/clip.py (hunk 5 above)
Comment (b0nce) [reply oml/models/vit/clip.py:1]: Ok, sure

File: oml/models/vit/clip.py (hunk 7 above)
Comment (b0nce) [reply oml/models/vit/clip.py:218]: This normalization is directly from the original paper. I thought it is much better to explicitly keep it the same.

File: oml/transforms/images/torchvision/transforms.py (hunk 1 above)
Comment (b0nce) [reply oml/transforms/images/torchvision/transforms.py:34]: Surprisingly, this way of preprocessing gives small increment when you try to check validation score on pretrained model.

File: oml/models/vit/clip.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            output_dim=embed_dim,
+        )
+
+        if weights is None:
+            return
+        if weights in self.pretrained_models:
+            weights, md5 = self.pretrained_models[weights]
+            weights = download_checkpoint(weights, md5)
+
+        try:  # check if weights are jitted
```
Comment (AlekseySh) [inline oml/models/vit/clip.py:93]: We already know which weights are jitted, so, instead of doing try/except, we can do 
```
JITTED_WEIGHTS = [name1, name2,...]
if weights in JITTED_WEIGHTS:
    do smth
```

File: oml/models/vit/clip.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    filter_state_dict,
+    patch_device_and_float,
+    remove_prefix_from_state_dict,
+)
+from oml.models.vit.vision_transformer_clip import VisionTransformer
+from oml.utils.io import download_checkpoint
+
+
+class ViTCLIPExtractor(IExtractor):
+    pretrained_models: Dict[str, Any] = {
```
Comment (AlekseySh) [inline oml/models/vit/clip.py:17]: can we squeeze the size of this dictionary like in ViTExtractor?
1. shorter the links moving prefix to the variable 
2. use only the first 6 letters of hash

thus, we may try to place everything in place

File: oml/models/vit/clip.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            weights = download_checkpoint(weights, md5)
+
+        try:  # check if weights are jitted
+            visual = torch.jit.load(Path(weights), map_location="cpu").visual.eval()
+            patch_device_and_float(visual, device="cpu")
+            state_dict = visual.state_dict()
+        except RuntimeError:
+            state_dict = torch.load(Path(weights), map_location="cpu")
+            state_dict = state_dict.get("state_dict", state_dict)
+            state_dict = remove_prefix_from_state_dict(state_dict, trial_key="conv1.weight", skip_unnecessary=True)
```
Comment (AlekseySh) [inline oml/models/vit/clip.py:100]: I was not sure about the need of this argument - `skip_unnecessary`
probably, after the refactoring suggested above we ca get rid of it

**Review by AlekseySh — COMMENTED** [review]
please, add a short import for the new clip model (put it to `oml.models.__init__`)

File: README.md [hunk 11]
```diff
@@ -354,6 +354,15 @@ model_from_disk = ViTExtractor(weights=oml.const.CKPT_SAVE_ROOT / "vits16_cars.c
 ```
 [comment]:checkpoint-end
 
+We also have some pretrained ViT-CLIP models with their zeros-shot scores (cmc@1) presented below:
```
Comment (AlekseySh) [inline README.md:357]: this is a table not only for clip, so, let's say `pretrained model` instead

Comment (AlekseySh) [reply README.md:357]: `zeros-shot` may be a confusing term. I suggest removing it:
`with their values of CMC@1 presented below:`

File: tests/test_oml/test_models/test_models_creation.py [hunk 12]
```diff
@@ -24,13 +30,21 @@ def test_creation(constructor: IExtractor, args: Dict[str, Any], default_arch: s
...
     for key in constructor.pretrained_models.keys():
-        net = constructor(weights=key, arch=key.split("_")[0], **args)
+        if constructor == ViTCLIPExtractor:
+            if not key.endswith("224"):
+                continue
+            arch = key.lstrip("sber_").lstrip("openai_")
+        else:
+            arch = key.split("_")[0]
+
+        print(constructor, key, arch, args)
```
[the remark this answers is not in the document]
Comment (b0nce) [reply tests/test_oml/test_models/test_models_creation.py:46]: Forget 😅

File: README.md (hunk 11 above)
Comment (b0nce) [reply README.md:357]: Actually, I'm not so sure about removing this because we are basically saying the same thing about models we have trained on these datasets. So if we leave only "pretrained ViT-CLIP models" in the sentence without explicitly stating that this is zero-shot performance, it may be confusing.
What do you think?

Comment (b0nce) [reply README.md:357]: Maybe we can say "CLIP-pretrained ViT models" 🤔

File: oml/models/vit/clip.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+            return
+        if weights in self.pretrained_models:
+            jitted_weights = weights in self.jitted_weights
+            weights, md5 = self.pretrained_models[weights]
+            weights = download_checkpoint(weights, md5)
+        else:
+            jitted_weights = False
+
+        if jitted_weights:  # check if weights are jitted
+            visual = torch.jit.load(Path(weights), map_location="cpu").visual.eval()
```
[the remark this answers is not in the document]
Comment (b0nce) [reply oml/models/vit/clip.py:94]: Surely, I've just copied this from my jupyter notebook where I needed this, thx.

File: README.md [hunk 14]
```diff
@@ -328,17 +328,40 @@ trainer.fit(pl_model, train_dataloaders=train_loader, val_dataloaders=val_loader
...
+| `ViTExtractor("vits16_dino", "vits16")`                   |                      0.629 |                0.456 |          0.693 |      0.313 |
+| `ViTExtractor("vits8_dino", "vits8")`                     |                      0.637 |                0.478 |          0.703 |      0.344 |
+| `ViTExtractor("vitb16_dino", "vitb16")`                   |                      0.636 |                0.464 |          0.626 |      0.340 |
+| `ViTExtractor("vitb8_dino", "vitb8")`                     |                      0.673 |                0.548 |          0.546 |      0.342 |
+| `ResnetExtractor("resnet50_moco_v2", "resnet50")`         |                      0.491 |                0.310 |          0.244 |      0.155 |
+
+*All figures above were obtained on the images with the sizes of 224 x 224.
 Note, that the models above expect the crop of the region of interest rather than the whole picture.
+It is also important to say that different models expect different preprocessing.
+For example, you should use `clip_transforms` for `ViTCLIPExtractor`*
```
Comment (AlekseySh) [inline README.md:361]: please, provide the full path to transforms like `oml.transforms...todo..clip_transforms`

File: README.md [hunk 15]
```diff
@@ -328,17 +328,40 @@ trainer.fit(pl_model, train_dataloaders=train_loader, val_dataloaders=val_loader
...
 |  `ViTExtractor(weights="vits16_sop", arch="vits16", ...)`   | 0.830 | Stanford Online Products | [link](https://drive.google.com/drive/folders/1WfPqCKbZ2KjRRQURGOOwrlQ87EUb7Zra?usp=sharing) | [link](https://github.com/OML-Team/open-metric-learning/tree/main/examples/sop/configs)      |        85cfa5        |
 |  `ViTExtractor(weights="vits16_cars", arch="vits16", ...)`  | 0.907 |         CARS 196         | [link](https://drive.google.com/drive/folders/17a4_fg94dox2sfkXmw-KCtiLBlx-ut-1?usp=sharing) | [link](https://github.com/OML-Team/open-metric-learning/tree/main/examples/cars/configs)     |        9f1e59        |
 |  `ViTExtractor(weights="vits16_cub", arch="vits16", ...)`   | 0.837 |       CUB 200 2011       | [link](https://drive.google.com/drive/folders/1TPCN-eZFLqoq4JBgnIfliJoEK48x9ozb?usp=sharing) | [link](https://github.com/OML-Team/open-metric-learning/tree/main/examples/cub/configs)      |        e82633        |
 
+We also provide an integration with the models pretrained by other researchers:
+
+|                        model                              |   Stanford Online Products |   DeepFashion InShop |   CUB 200 2011 |   CARS 196 |
```
Comment (AlekseySh) [inline README.md:344]: `model` head is not centered

File: oml/transforms/images/albumentations/transforms.py [hunk 16]
```diff
@@ -136,6 +136,12 @@ def get_normalisation_resize_albu(im_size: int, mean: TNormParam = MEAN, std: TN
     )
 
 
+def get_clip_transforms(im_size: int) -> albu.Compose:
```
Comment (AlekseySh) [inline oml/transforms/images/albumentations/transforms.py:139]: let's follow our existing naming schedule for transforms
so, it this case it would be `get_normalisation_resize_albu_clip`

Comment (AlekseySh) [reply oml/transforms/images/albumentations/transforms.py:139]: btw, don't forget to add it to __all__

File: oml/transforms/images/albumentations/transforms.py [hunk 17]
```diff
@@ -136,6 +136,12 @@ def get_normalisation_resize_albu(im_size: int, mean: TNormParam = MEAN, std: TN
...
+    return get_normalisation_resize_albu(
+        im_size=im_size, mean=(0.48145466, 0.4578275, 0.40821073), std=(0.26862954, 0.26130258, 0.27577711)
```
Comment (AlekseySh) [inline oml/transforms/images/albumentations/transforms.py:141]: let's move these values as constants to const.py:
`MEAN_CLIP`, `STD_CLIP`

**Review by AlekseySh — CHANGES_REQUESTED** [review]

**Review by AlekseySh — APPROVED** [review]
